fairmq/shmem/FairMQSocketSHM.cxx

```diff
@@ -210,151 +220,150 @@ int FairMQSocketSHM::Receive(FairMQMessagePtr& msg, const int flags)
+            if (!fInterrupted && ((flags & ZMQ_DONTWAIT) == 0))
+            {
+                continue;
+            }
+            else
+            {
+                zmq_msg_close (&lZmqMsg);
+                return -2;
+            }
+        }
+        else if (zmq_errno() == ETERM)
+        {
+            zmq_msg_close (&lZmqMsg);
+            LOG(INFO) << "terminating socket " << fId;
+            return -1;
+        }
+        else
+        {
+            zmq_msg_close (&lZmqMsg);
+            LOG(ERROR) << "Failed sending on socket " << fId << ", reason: " << zmq_strerror(errno);
+            return nbytes;
+        }
     }
 }
 
+
 int64_t FairMQSocketSHM::Receive(vector<FairMQMessagePtr>& msgVec, const int flags)
 {
     int64_t totalSize = 0;
-    int64_t more = 0;
-    bool repeat = false;
 
-    while (true)
+    while (!fInterrupted)
     {
-        // Warn if the vector is filled before Receive() and empty it.
-        // if (msgVec.size() > 0)
-        // {
-        //     LOG(warn) << "Message vector contains elements before Receive(), they will be deleted!";
-        //     msgVec.clear();
-        // }
+        zmq_msg_t lRcvMsg;
+        zmq_msg_init(&lRcvMsg);
+        int nbytes = zmq_msg_recv(&lRcvMsg, fSocket, flags);
+        if (nbytes == 0)
+        {
+            zmq_msg_close (&lRcvMsg);
+            return 0;
+        }
+        else if (nbytes > 0)
+        {
+            MetaHeader* lHdrVec = static_cast<MetaHeader*>(zmq_msg_data(&lRcvMsg));
+            const auto lHdrVecSize = zmq_msg_size(&lRcvMsg);
+            assert(lHdrVecSize > 0);
+            assert(lHdrVecSize % sizeof(MetaHeader) == 0);
 
-        totalSize = 0;
-        more = 0;
-        repeat = false;
+            const auto lNumMessages = lHdrVecSize / sizeof (MetaHeader);
 
-        do
-        {
-            FairMQMessagePtr part(new FairMQMessageSHM(fManager));
-            zmq_msg_t* msgPtr = static_cast<FairMQMessageSHM*>(part.get())->GetMessage();
+            msgVec.reserve(lNumMessages);
 
-            int nbytes = zmq_msg_recv(msgPtr, fSocket, flags);
-            if (nbytes == 0)
+            for (auto m = 0; m < lNumMessages; m++)
             {
-                msgVec.push_back(move(part));
-            }
-            else if (nbytes > 0)
-            {
-                MetaHeader* hdr = static_cast<MetaHeader*>(zmq_msg_data(msgPtr));
-                size_t size = 0;
-                static_cast<FairMQMessageSHM*>(part.get())->fHandle = hdr->fHandle;
-                static_cast<FairMQMessageSHM*>(part.get())->fSize = hdr->fSize;
-                static_cast<FairMQMessageSHM*>(part.get())->fRegionId = hdr->fRegionId;
-                static_cast<FairMQMessageSHM*>(part.get())->fHint = hdr->fHint;
-                size = part->GetSize();
+                MetaHeader lMetaHeader;
+                memcpy(&lMetaHeader, &lHdrVec[m], sizeof(MetaHeader));
+
+                msgVec.emplace_back(std::make_unique<FairMQMessageSHM>(fManager));
```

Comment:
Could you replace std::make_unique with fair::mq::tools::make_unique, we still do not require c++14

Reply:
sure, branch updated